jrubyJar task failing on missing dependency to Apache zip

3:25:30 PM: Executing task 'jrubyJar'...

:prepareJRubyJar UP-TO-DATE
:jrubyJar FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':jrubyJar'.
> org/apache/tools/zip/ZipOutputStream

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 0.238 secs
org/apache/tools/zip/ZipOutputStream
3:25:30 PM: Task execution finished 'jrubyJar'.

Which version of Gradle & JDK?


Sent from my Samsung device

-------- Original message --------
From: amiracam <notifications@github.com> 
Date: 26/09/2017  21:33  (GMT+01:00) 
To: jruby-gradle/jruby-gradle-plugin <jruby-gradle-plugin@noreply.github.com> 
Cc: Subscribed <subscribed@noreply.github.com> 
Subject: [jruby-gradle/jruby-gradle-plugin] jrubyJar task failing on missing
  dependency to Apache zip (#307) 

3:25:30 PM: Executing task 'jrubyJar'...
:prepareJRubyJar UP-TO-DATE

:jrubyJar FAILED
FAILURE: Build failed with an exception.

What went wrong:

Execution failed for task ':jrubyJar'.


org/apache/tools/zip/ZipOutputStream


Try:

Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED
Total time: 0.238 secs

org/apache/tools/zip/ZipOutputStream

3:25:30 PM: Task execution finished 'jrubyJar'.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/jruby-gradle/jruby-gradle-plugin","title":"jruby-gradle/jruby-gradle-plugin","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/jruby-gradle/jruby-gradle-plugin"}},"updates":{"snippets":[{"icon":"DESCRIPTION","message":"jrubyJar task failing on missing dependency to Apache zip (#307)"}],"action":{"name":"View Issue","url":"https://github.com/jruby-gradle/jruby-gradle-plugin/issues/307"}}}

Gradle 4.2 , Jdk 8.121 , 131

On Tue, Sep 26, 2017 at 4:10 PM Schalk W. Cronjé <notifications@github.com>
wrote:

>
>
> Which version of Gradle & JDK?
>
>
> Sent from my Samsung device
>
> -------- Original message --------
> From: amiracam <notifications@github.com>
> Date: 26/09/2017 21:33 (GMT+01:00)
> To: jruby-gradle/jruby-gradle-plugin <
> jruby-gradle-plugin@noreply.github.com>
> Cc: Subscribed <subscribed@noreply.github.com>
> Subject: [jruby-gradle/jruby-gradle-plugin] jrubyJar task failing on
> missing dependency to Apache zip (#307)
>
> 3:25:30 PM: Executing task 'jrubyJar'...
> :prepareJRubyJar UP-TO-DATE
>
> :jrubyJar FAILED
> FAILURE: Build failed with an exception.
>
> What went wrong:
>
> Execution failed for task ':jrubyJar'.
>
>
> org/apache/tools/zip/ZipOutputStream
>
>
> Try:
>
> Run with --stacktrace option to get the stack trace. Run with --info or
> --debug option to get more log output.
>
> BUILD FAILED
> Total time: 0.238 secs
>
> org/apache/tools/zip/ZipOutputStream
>
> 3:25:30 PM: Task execution finished 'jrubyJar'.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
>
>
>
>
>
>
>
> {"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/jruby-gradle/jruby-gradle-plugin","title":"jruby-gradle/jruby-gradle-plugin","subtitle":"GitHub
> repository","main_image_url":"
> https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png
> ","avatar_image_url":"
> https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open
> in GitHub","url":"https://github.com/jruby-gradle/jruby-gradle-plugin"}},"updates":{"snippets":[{"icon":"DESCRIPTION","message":"jrubyJar
> task failing on missing dependency to Apache zip
> (#307)"}],"action":{"name":"View Issue","url":"
> https://github.com/jruby-gradle/jruby-gradle-plugin/issues/307"}}}
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jruby-gradle/jruby-gradle-plugin/issues/307#issuecomment-332320414>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAfB-L5tCEYvx9rTxlLsVZPytUfcNxfeks5smVomgaJpZM4PkxFE>
> .
>
-- 
Charles A. Monteiro
www.monteirosfusion.com
sent from the road

--- a/jar-plugin/build.gradle
+++ b/jar-plugin/build.gradle
@@ -17,7 +17,14 @@ generateTestConfig {
 
 dependencies {
     compile project(':jruby-gradle-plugin')
-    compile 'com.github.jengelman.gradle.plugins:shadow:[1.2.2,2.0)'  // NEED TO FIX THIS OPEN END
+    /*
+     * NOTE: version 5.0.0 of the shadow plugin supports only Gradle 5.x and later
+     */
+    compile 'com.github.jengelman.gradle.plugins:shadow:[4.0.2,5.0)'
+    compile 'org.codehaus.plexus:plexus-utils:[3.2.0,3.3)'
+    compile 'org.apache.commons:commons-io:1.3.2'
+    compile 'org.ow2.asm:asm-commons:[6.1,6.99)'
+    compile 'org.apache.ant:ant:[1.10.6,2.0)'
 
     testCompile (spockVersion) {
         exclude module : 'groovy-all'

--- a/jar-plugin/build.gradle
+++ b/jar-plugin/build.gradle
@@ -20,7 +20,7 @@ dependencies {
     /*
      * NOTE: version 5.0.0 of the shadow plugin supports only Gradle 5.x and later
      */
-    compile 'com.github.jengelman.gradle.plugins:shadow:[4.0.2,5.0)'
+    compile 'com.github.jengelman.gradle.plugins:shadow:4.0.4'
     compile 'org.codehaus.plexus:plexus-utils:[3.2.0,3.3)'
     compile 'org.apache.commons:commons-io:1.3.2'
     compile 'org.ow2.asm:asm-commons:[6.1,6.99)'

--- a/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/JRubyJar.groovy
+++ b/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/JRubyJar.groovy
@@ -1,21 +1,20 @@
 package com.github.jrubygradle.jar
 
-import com.github.jengelman.gradle.plugins.shadow.internal.DefaultZipCompressor
+import com.github.jengelman.gradle.plugins.shadow.internal.GradleVersionUtil
 import com.github.jengelman.gradle.plugins.shadow.internal.ZipCompressor
 import com.github.jrubygradle.JRubyPrepare
 import com.github.jrubygradle.jar.internal.JRubyDirInfoTransformer
 import com.github.jrubygradle.jar.internal.JRubyJarCopyAction
 import groovy.transform.PackageScope
-import org.apache.tools.zip.ZipOutputStream
 import org.gradle.api.InvalidUserDataException
 import org.gradle.api.artifacts.Configuration
 import org.gradle.api.file.DuplicatesStrategy
 import org.gradle.api.internal.file.copy.CopyAction
 import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.Internal
 import org.gradle.api.tasks.Optional
 import org.gradle.api.tasks.StopExecutionException
 import org.gradle.api.tasks.bundling.Jar
-import org.gradle.api.tasks.bundling.ZipEntryCompression
 
 /**
  * JRubyJar creates a Java Archive with Ruby code packed inside of it.
@@ -241,6 +240,7 @@ class JRubyJar extends Jar {
         appendix = 'jruby'
         /* Make sure our default configuration is present regardless of whether we use it or not */
         prepareTask = project.task("prepare${prepareNameForSuffix(name)}", type: JRubyPrepare)
+        versionUtil = new GradleVersionUtil(getProject().getGradle().getGradleVersion())
         dependsOn prepareTask
 
         // TODO get rid of this and try to adjust the CopySpec for the gems
@@ -293,20 +293,21 @@ class JRubyJar extends Jar {
         return new JRubyJarCopyAction(getArchivePath(),
                 getInternalCompressor(),
                 null, /* DocumentationRegistry */
+                'utf-8', /* encoding */
                 [new JRubyDirInfoTransformer()], /* transformers */
                 [], /* relocators */
-                mainSpec.buildRootResolver().getPatternSet())
+                mainSpec.buildRootResolver().getPatternSet(), /* patternSet */
+                versionUtil, /* util */
+                false, /* preserveFileTimestamps */
+                false, /* minimizeJar */
+                null /* unusedTracker */
+                )
+
     }
 
+    @Internal
     protected ZipCompressor getInternalCompressor() {
-        switch (entryCompression) {
-            case ZipEntryCompression.DEFLATED:
-                return new DefaultZipCompressor(this.zip64, ZipOutputStream.DEFLATED)
-            case ZipEntryCompression.STORED:
-                return new DefaultZipCompressor(this.zip64, ZipOutputStream.STORED)
-            default:
-                throw new IllegalArgumentException(String.format('Unknown Compression type %s', entryCompression))
-        }
+        return versionUtil.getInternalCompressor(getEntryCompression(), this)
     }
 
     /**
@@ -328,4 +329,5 @@ class JRubyJar extends Jar {
     protected String embeddedJRubyMainsVersion = DEFAULT_JRUBY_MAINS
     protected String jarConfiguration = DEFAULT_JRUBYJAR_CONFIG
     protected String jarMainClass
+    protected GradleVersionUtil versionUtil
 }

--- a/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/JRubyJar.groovy
+++ b/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/JRubyJar.groovy
@@ -1,11 +1,17 @@
 package com.github.jrubygradle.jar
 
-import com.github.jengelman.gradle.plugins.shadow.internal.GradleVersionUtil
+/*
+ * These two internal imports from the Shadow plugin are unavoidable because of
+ * the expected internals of ShadowCopyAction
+ */
+import com.github.jengelman.gradle.plugins.shadow.internal.DefaultZipCompressor
 import com.github.jengelman.gradle.plugins.shadow.internal.ZipCompressor
+
 import com.github.jrubygradle.JRubyPrepare
 import com.github.jrubygradle.jar.internal.JRubyDirInfoTransformer
 import com.github.jrubygradle.jar.internal.JRubyJarCopyAction
 import groovy.transform.PackageScope
+import org.apache.tools.zip.ZipOutputStream
 import org.gradle.api.InvalidUserDataException
 import org.gradle.api.artifacts.Configuration
 import org.gradle.api.file.DuplicatesStrategy
@@ -15,6 +21,7 @@ import org.gradle.api.tasks.Internal
 import org.gradle.api.tasks.Optional
 import org.gradle.api.tasks.StopExecutionException
 import org.gradle.api.tasks.bundling.Jar
+import org.gradle.api.tasks.bundling.ZipEntryCompression
 
 /**
  * JRubyJar creates a Java Archive with Ruby code packed inside of it.
@@ -240,7 +247,6 @@ class JRubyJar extends Jar {
         appendix = 'jruby'
         /* Make sure our default configuration is present regardless of whether we use it or not */
         prepareTask = project.task("prepare${prepareNameForSuffix(name)}", type: JRubyPrepare)
-        versionUtil = new GradleVersionUtil(getProject().getGradle().getGradleVersion())
         dependsOn prepareTask
 
         // TODO get rid of this and try to adjust the CopySpec for the gems
@@ -297,7 +303,6 @@ class JRubyJar extends Jar {
                 [new JRubyDirInfoTransformer()], /* transformers */
                 [], /* relocators */
                 mainSpec.buildRootResolver().getPatternSet(), /* patternSet */
-                versionUtil, /* util */
                 false, /* preserveFileTimestamps */
                 false, /* minimizeJar */
                 null /* unusedTracker */
@@ -307,7 +312,14 @@ class JRubyJar extends Jar {
 
     @Internal
     protected ZipCompressor getInternalCompressor() {
-        return versionUtil.getInternalCompressor(getEntryCompression(), this)
+        switch (entryCompression) {
+            case ZipEntryCompression.DEFLATED:
+                return new DefaultZipCompressor(this.zip64, ZipOutputStream.DEFLATED)
+            case ZipEntryCompression.STORED:
+                return new DefaultZipCompressor(this.zip64, ZipOutputStream.STORED)
+            default:
+                throw new IllegalArgumentException(String.format('Unknown Compression type %s', entryCompression))
+        }
     }
 
     /**
@@ -329,5 +341,4 @@ class JRubyJar extends Jar {
     protected String embeddedJRubyMainsVersion = DEFAULT_JRUBY_MAINS
     protected String jarConfiguration = DEFAULT_JRUBYJAR_CONFIG
     protected String jarMainClass
-    protected GradleVersionUtil versionUtil
 }

--- a/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/internal/JRubyDirInfoTransformer.groovy
+++ b/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/internal/JRubyDirInfoTransformer.groovy
@@ -4,13 +4,13 @@ import java.nio.file.Files
 import java.nio.file.Path
 import java.nio.file.Paths
 
-import org.apache.tools.zip.ZipOutputStream
 import org.apache.tools.zip.ZipEntry
 import org.codehaus.plexus.util.IOUtil
 import org.gradle.api.file.FileTreeElement
 
-import com.github.jengelman.gradle.plugins.shadow.relocation.Relocator
+import shadow.org.apache.tools.zip.ZipOutputStream
 import com.github.jengelman.gradle.plugins.shadow.transformers.Transformer
+import com.github.jengelman.gradle.plugins.shadow.transformers.TransformerContext
 
 /**
  * JRubyDirInfoTransformer implements a {@link Transformer} interface.
@@ -44,7 +44,7 @@ class JRubyDirInfoTransformer implements Transformer {
     }
 
     /** No-op since we don't transform the actual file */
-    void transform(String path, InputStream is, List<Relocator> relocators) {
+    void transform(TransformerContext context) {
         return
     }
 
@@ -63,7 +63,7 @@ class JRubyDirInfoTransformer implements Transformer {
      * This method will also clean up our tempdir to make sure we don't
      * clutter the user's machine with junk
      */
-    void modifyOutputStream(ZipOutputStream os) {
+    void modifyOutputStream(ZipOutputStream os, boolean preserveFileTimestamps) {
         processDirectory(os, tmpDir)
         deleteTempDirectory(tmpDir)
     }

--- a/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/internal/JRubyDirInfoTransformer.groovy
+++ b/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/internal/JRubyDirInfoTransformer.groovy
@@ -4,7 +4,7 @@ import java.nio.file.Files
 import java.nio.file.Path
 import java.nio.file.Paths
 
-import org.apache.tools.zip.ZipEntry
+import shadow.org.apache.tools.zip.ZipEntry
 import org.codehaus.plexus.util.IOUtil
 import org.gradle.api.file.FileTreeElement
 

--- a/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/internal/JRubyJarCopyAction.groovy
+++ b/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/internal/JRubyJarCopyAction.groovy
@@ -32,7 +32,6 @@ import org.gradle.api.internal.file.DefaultFileTreeElement
 import org.gradle.api.internal.file.copy.CopyAction
 import org.gradle.api.internal.file.copy.CopyActionProcessingStream
 import org.gradle.api.internal.file.copy.FileCopyDetailsInternal
-import org.gradle.api.specs.Spec
 import org.gradle.api.tasks.WorkResult
 import org.gradle.api.tasks.bundling.Zip
 import org.gradle.api.tasks.util.PatternSet
@@ -238,81 +237,26 @@ class JRubyJarCopyAction implements CopyAction {
 
         @Override
         void visitFile(FileCopyDetails fileDetails) {
-            if (!isArchive(fileDetails)) {
-                try {
-                    boolean isClass = isClass(fileDetails)
-                    if (!remapper.hasRelocators() || !isClass) {
-                        if (!isTransformable(fileDetails)) {
-                            String mappedPath = remapper.map(fileDetails.relativePath.pathString)
-                            ZipEntry archiveEntry = new ZipEntry(mappedPath)
-                            archiveEntry.setTime(getArchiveTimeFor(fileDetails.lastModified))
-                            archiveEntry.unixMode = (UnixStat.FILE_FLAG | fileDetails.mode)
-                            zipOutStr.putNextEntry(archiveEntry)
-                            fileDetails.copyTo(zipOutStr)
-                            zipOutStr.closeEntry()
-                        } else {
-                            transform(fileDetails)
-                        }
-                    } else if (isClass && !isUnused(fileDetails.path)) {
-                        remapClass(fileDetails)
-                    }
-                    recordVisit(fileDetails.relativePath)
-                } catch (Exception e) {
-                    throw new GradleException(String.format("Could not add %s to ZIP '%s'.", fileDetails, zipFile), e)
-                }
-            } else {
-                processArchive(fileDetails)
-            }
-        }
-
-        private void processArchive(FileCopyDetails fileDetails) {
-            ZipFile archive = new ZipFile(fileDetails.file)
             try {
-                List<ArchiveFileTreeElement> archiveElements = archive.entries.collect {
-                    new ArchiveFileTreeElement(new RelativeArchivePath(it))
-                }
-                Spec<FileTreeElement> patternSpec = patternSet.asSpec
-                List<ArchiveFileTreeElement> filteredArchiveElements = archiveElements.findAll { ArchiveFileTreeElement archiveElement ->
-                    patternSpec.isSatisfiedBy(archiveElement.asFileTreeElement())
-                }
-                filteredArchiveElements.each { ArchiveFileTreeElement archiveElement ->
-                    if (archiveElement.relativePath.file) {
-                        visitArchiveFile(archiveElement, archive)
-                    }
-                }
-            } finally {
-                archive.close()
-            }
-        }
-
-        private void visitArchiveDirectory(RelativeArchivePath archiveDir) {
-            if (recordVisit(archiveDir)) {
-                zipOutStr.putNextEntry(archiveDir.entry)
-                zipOutStr.closeEntry()
-            }
-        }
-
-        private void visitArchiveFile(ArchiveFileTreeElement archiveFile, ZipFile archive) {
-            RelativeArchivePath archiveFilePath = archiveFile.relativePath
-            if (archiveFile.classFile || !isTransformable(archiveFile)) {
-                if (recordVisit(archiveFilePath) && !isUnused(archiveFilePath.entry.name)) {
-                    if (!remapper.hasRelocators() || !archiveFile.classFile) {
-                        copyArchiveEntry(archiveFilePath, archive)
+                boolean isClass = isClass(fileDetails)
+                if (!remapper.hasRelocators() || !isClass) {
+                    if (!isTransformable(fileDetails)) {
+                        String mappedPath = remapper.map(fileDetails.relativePath.pathString)
+                        ZipEntry archiveEntry = new ZipEntry(mappedPath)
+                        archiveEntry.setTime(getArchiveTimeFor(fileDetails.lastModified))
+                        archiveEntry.unixMode = (UnixStat.FILE_FLAG | fileDetails.mode)
+                        zipOutStr.putNextEntry(archiveEntry)
+                        fileDetails.copyTo(zipOutStr)
+                        zipOutStr.closeEntry()
                     } else {
-                        remapClass(archiveFilePath, archive)
+                        transform(fileDetails)
                     }
+                } else if (isClass && !isUnused(fileDetails.path)) {
+                    remapClass(fileDetails)
                 }
-            } else {
-                transform(archiveFile, archive)
-            }
-        }
-
-        private void addParentDirectories(RelativeArchivePath file) {
-            if (file) {
-                addParentDirectories(file.parent)
-                if (!file.file) {
-                    visitArchiveDirectory(file)
-                }
+                recordVisit(fileDetails.relativePath)
+            } catch (Exception e) {
+                throw new GradleException(String.format("Could not add %s to ZIP '%s'.", fileDetails, zipFile), e)
             }
         }
 
@@ -382,22 +326,6 @@ class JRubyJarCopyAction implements CopyAction {
             } finally {
                 bis.close()
             }
-        }
-
-        private void copyArchiveEntry(RelativeArchivePath archiveFile, ZipFile archive) {
-            String mappedPath = remapper.map(archiveFile.entry.name)
-            ZipEntry entry = new ZipEntry(mappedPath)
-            entry.setTime(getArchiveTimeFor(archiveFile.entry.time))
-            RelativeArchivePath mappedFile = new RelativeArchivePath(entry)
-            addParentDirectories(mappedFile)
-            zipOutStr.putNextEntry(mappedFile.entry)
-            InputStream is = archive.getInputStream(archiveFile.entry)
-            try {
-                IOUtils.copyLarge(is, zipOutStr)
-            } finally {
-                is.close()
-            }
-            zipOutStr.closeEntry()
         }
 
         @Override

--- a/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/internal/JRubyJarCopyAction.groovy
+++ b/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/internal/JRubyJarCopyAction.groovy
@@ -7,7 +7,6 @@ package com.github.jrubygradle.jar.internal
 */
 
 import com.github.jengelman.gradle.plugins.shadow.impl.RelocatorRemapper
-import com.github.jengelman.gradle.plugins.shadow.internal.GradleVersionUtil
 import com.github.jengelman.gradle.plugins.shadow.internal.UnusedTracker
 import com.github.jengelman.gradle.plugins.shadow.internal.ZipCompressor
 import com.github.jengelman.gradle.plugins.shadow.relocation.Relocator
@@ -33,6 +32,7 @@ import org.gradle.api.internal.file.copy.CopyAction
 import org.gradle.api.internal.file.copy.CopyActionProcessingStream
 import org.gradle.api.internal.file.copy.FileCopyDetailsInternal
 import org.gradle.api.tasks.WorkResult
+import org.gradle.api.tasks.WorkResults
 import org.gradle.api.tasks.bundling.Zip
 import org.gradle.api.tasks.util.PatternSet
 import org.gradle.internal.UncheckedException
@@ -69,14 +69,13 @@ class JRubyJarCopyAction implements CopyAction {
     private final List<Relocator> relocators
     private final PatternSet patternSet
     private final String encoding
-    private final GradleVersionUtil versionUtil
     private final boolean preserveFileTimestamps
     private final boolean minimizeJar
     private final UnusedTracker unusedTracker
 
     JRubyJarCopyAction(File zipFile, ZipCompressor compressor, DocumentationRegistry documentationRegistry,
                             String encoding, List<Transformer> transformers, List<Relocator> relocators,
-                            PatternSet patternSet, GradleVersionUtil util,
+                            PatternSet patternSet,
                             boolean preserveFileTimestamps, boolean minimizeJar, UnusedTracker unusedTracker) {
 
         this.zipFile = zipFile
@@ -86,7 +85,6 @@ class JRubyJarCopyAction implements CopyAction {
         this.relocators = relocators
         this.patternSet = patternSet
         this.encoding = encoding
-        this.versionUtil = util
         this.preserveFileTimestamps = preserveFileTimestamps
         this.minimizeJar = minimizeJar
         this.unusedTracker = unusedTracker
@@ -141,7 +139,7 @@ class JRubyJarCopyAction implements CopyAction {
                 )
             }
         }
-        return versionUtil.getWorkResult(true)
+        return WorkResults.didWork(true)
     }
 
     private void processTransformers(ZipOutputStream stream) {

--- a/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/internal/JRubyJarCopyAction.groovy
+++ b/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/internal/JRubyJarCopyAction.groovy
@@ -37,7 +37,7 @@ import org.gradle.internal.UncheckedException
 import org.objectweb.asm.ClassReader
 import org.objectweb.asm.ClassVisitor
 import org.objectweb.asm.ClassWriter
-import org.objectweb.asm.commons.RemappingClassAdapter
+import org.objectweb.asm.commons.ClassRemapper
 
 import java.util.zip.ZipException
 
@@ -235,7 +235,7 @@ class JRubyJarCopyAction implements CopyAction {
             // that use the constant pool to determine the dependencies of a class.
             ClassWriter cw = new ClassWriter(0)
 
-            ClassVisitor cv = new RemappingClassAdapter(cw, remapper)
+            ClassVisitor cv = new ClassRemapper(cw, remapper)
 
             try {
                 cr.accept(cv, ClassReader.EXPAND_FRAMES)

--- a/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/internal/JRubyJarCopyAction.groovy
+++ b/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/internal/JRubyJarCopyAction.groovy
@@ -7,17 +7,19 @@ package com.github.jrubygradle.jar.internal
 */
 
 import com.github.jengelman.gradle.plugins.shadow.impl.RelocatorRemapper
+import com.github.jengelman.gradle.plugins.shadow.internal.GradleVersionUtil
+import com.github.jengelman.gradle.plugins.shadow.internal.UnusedTracker
 import com.github.jengelman.gradle.plugins.shadow.internal.ZipCompressor
 import com.github.jengelman.gradle.plugins.shadow.relocation.Relocator
 import com.github.jengelman.gradle.plugins.shadow.transformers.Transformer
 import groovy.util.logging.Slf4j
 import org.apache.commons.io.FilenameUtils
 import org.apache.commons.io.IOUtils
-import org.apache.tools.zip.UnixStat
-import org.apache.tools.zip.Zip64RequiredException
-import org.apache.tools.zip.ZipEntry
-import org.apache.tools.zip.ZipFile
-import org.apache.tools.zip.ZipOutputStream
+import shadow.org.apache.tools.zip.UnixStat
+import shadow.org.apache.tools.zip.Zip64RequiredException
+import shadow.org.apache.tools.zip.ZipEntry
+import shadow.org.apache.tools.zip.ZipFile
+import shadow.org.apache.tools.zip.ZipOutputStream
 import org.gradle.api.Action
 import org.gradle.api.GradleException
 import org.gradle.api.UncheckedIOException
@@ -26,10 +28,11 @@ import org.gradle.api.file.FileTreeElement
 import org.gradle.api.file.RelativePath
 import org.gradle.api.internal.DocumentationRegistry
 import org.gradle.api.internal.file.CopyActionProcessingStreamAction
+import org.gradle.api.internal.file.DefaultFileTreeElement
 import org.gradle.api.internal.file.copy.CopyAction
 import org.gradle.api.internal.file.copy.CopyActionProcessingStream
 import org.gradle.api.internal.file.copy.FileCopyDetailsInternal
-import org.gradle.api.tasks.WorkResults
+import org.gradle.api.specs.Spec
 import org.gradle.api.tasks.WorkResult
 import org.gradle.api.tasks.bundling.Zip
 import org.gradle.api.tasks.util.PatternSet
@@ -46,7 +49,7 @@ import java.util.zip.ZipException
  *
  * This class, in its current form is really just a big copy and paste of the
  * shadow plugin's <a
- * href="https://github.com/johnrengelman/shadow/blob/51f2b5916edd7690dca5e89e8b3a8f330d435423/src/main/groovy/com/github/jengelman/gradle/plugins/shadow/tasks/ShadowCopyAction.groovy">ShadowCopyAction</a>.
+ * href="https://github.com/johnrengelman/shadow/blob/4.0.4/src/main/groovy/com/github/jengelman/gradle/plugins/shadow/tasks/ShadowCopyAction.groovy">ShadowCopyAction</a>
  * with one notable exception, it disables the behavior of unzipping nested
  * archives when creating the resulting archive.
  *
@@ -55,8 +58,10 @@ import java.util.zip.ZipException
  * support the same behavior in a less hackish way.
  */
 @Slf4j
-class ShadowCopyAction implements CopyAction {
-    static final long CONSTANT_TIME_FOR_ZIP_ENTRIES = (new GregorianCalendar(1980, 1, 1, 0, 0, 0)).getTimeInMillis()
+@SuppressWarnings(['ParameterCount', 'CatchException', 'DuplicateStringLiteral',
+'CatchThrowable', 'VariableName', 'UnnecessaryGString', 'InvertedIfElse'])
+class JRubyJarCopyAction implements CopyAction {
+    static final long CONSTANT_TIME_FOR_ZIP_ENTRIES = (new GregorianCalendar(1980, 1, 1, 0, 0, 0)).timeInMillis
 
     private final File zipFile
     private final ZipCompressor compressor
@@ -64,16 +69,15 @@ class ShadowCopyAction implements CopyAction {
     private final List<Transformer> transformers
     private final List<Relocator> relocators
     private final PatternSet patternSet
-    private final ShadowStats stats
     private final String encoding
     private final GradleVersionUtil versionUtil
     private final boolean preserveFileTimestamps
     private final boolean minimizeJar
     private final UnusedTracker unusedTracker
 
-    ShadowCopyAction(File zipFile, ZipCompressor compressor, DocumentationRegistry documentationRegistry,
+    JRubyJarCopyAction(File zipFile, ZipCompressor compressor, DocumentationRegistry documentationRegistry,
                             String encoding, List<Transformer> transformers, List<Relocator> relocators,
-                            PatternSet patternSet, ShadowStats stats, GradleVersionUtil util,
+                            PatternSet patternSet, GradleVersionUtil util,
                             boolean preserveFileTimestamps, boolean minimizeJar, UnusedTracker unusedTracker) {
 
         this.zipFile = zipFile
@@ -82,7 +86,6 @@ class ShadowCopyAction implements CopyAction {
         this.transformers = transformers
         this.relocators = relocators
         this.patternSet = patternSet
-        this.stats = stats
         this.encoding = encoding
         this.versionUtil = util
         this.preserveFileTimestamps = preserveFileTimestamps
@@ -122,7 +125,7 @@ class ShadowCopyAction implements CopyAction {
                 void execute(ZipOutputStream outputStream) {
                     try {
                         stream.process(new StreamAction(outputStream, encoding, transformers, relocators, patternSet,
-                                unusedClasses, stats))
+                                unusedClasses))
                         processTransformers(outputStream)
                     } catch (Exception e) {
                         log.error('ex', e)
@@ -164,11 +167,11 @@ class ShadowCopyAction implements CopyAction {
     private static <T extends Closeable> void withResource(T resource, Action<? super T> action) {
         try {
             action.execute(resource)
-        } catch(Throwable t) {
+        } catch (Throwable t) {
             try {
                 resource.close()
             } catch (IOException e) {
-                // Ignored
+                log.debug("Dropping ignored exception ${e}")
             }
             throw UncheckedException.throwAsUncheckedException(t)
         }
@@ -198,7 +201,9 @@ class ShadowCopyAction implements CopyAction {
             }
         }
 
-        protected void visitDir(FileCopyDetails dirDetails) {}
+        @SuppressWarnings(['UnusedMethodParameter', 'EmptyMethodInAbstractClass'])
+        protected void visitDir(FileCopyDetails dirDetails) {
+        }
 
         protected abstract void visitFile(FileCopyDetails fileDetails)
     }
@@ -211,21 +216,18 @@ class ShadowCopyAction implements CopyAction {
         private final RelocatorRemapper remapper
         private final PatternSet patternSet
         private final Set<String> unused
-        private final ShadowStats stats
 
-        private Set<String> visitedFiles = new HashSet<String>()
+        private final Set<String> visitedFiles = [] as Set
 
         StreamAction(ZipOutputStream zipOutStr, String encoding, List<Transformer> transformers,
-                            List<Relocator> relocators, PatternSet patternSet, Set<String> unused,
-                            ShadowStats stats) {
+                            List<Relocator> relocators, PatternSet patternSet, Set<String> unused) {
             this.zipOutStr = zipOutStr
             this.transformers = transformers
             this.relocators = relocators
-            this.remapper = new RelocatorRemapper(relocators, stats)
+            this.remapper = new RelocatorRemapper(relocators, null)
             this.patternSet = patternSet
             this.unused = unused
-            this.stats = stats
-            if(encoding != null) {
+            if (encoding != null) {
                 this.zipOutStr.setEncoding(encoding)
             }
         }
@@ -264,13 +266,12 @@ class ShadowCopyAction implements CopyAction {
         }
 
         private void processArchive(FileCopyDetails fileDetails) {
-            stats.startJar()
             ZipFile archive = new ZipFile(fileDetails.file)
             try {
                 List<ArchiveFileTreeElement> archiveElements = archive.entries.collect {
                     new ArchiveFileTreeElement(new RelativeArchivePath(it))
                 }
-                Spec<FileTreeElement> patternSpec = patternSet.getAsSpec()
+                Spec<FileTreeElement> patternSpec = patternSet.asSpec
                 List<ArchiveFileTreeElement> filteredArchiveElements = archiveElements.findAll { ArchiveFileTreeElement archiveElement ->
                     patternSpec.isSatisfiedBy(archiveElement.asFileTreeElement())
                 }
@@ -282,7 +283,6 @@ class ShadowCopyAction implements CopyAction {
             } finally {
                 archive.close()
             }
-            stats.finishJar()
         }
 
         private void visitArchiveDirectory(RelativeArchivePath archiveDir) {
@@ -293,7 +293,7 @@ class ShadowCopyAction implements CopyAction {
         }
 
         private void visitArchiveFile(ArchiveFileTreeElement archiveFile, ZipFile archive) {
-            def archiveFilePath = archiveFile.relativePath
+            RelativeArchivePath archiveFilePath = archiveFile.relativePath
             if (archiveFile.classFile || !isTransformable(archiveFile)) {
                 if (recordVisit(archiveFilePath) && !isUnused(archiveFilePath.entry.name)) {
                     if (!remapper.hasRelocators() || !archiveFile.classFile) {
@@ -432,7 +432,6 @@ class ShadowCopyAction implements CopyAction {
                                 .path(mappedPath)
                                 .is(is)
                                 .relocators(relocators)
-                                .stats(stats)
                                 .build()
                 )
             } finally {
@@ -462,11 +461,10 @@ class ShadowCopyAction implements CopyAction {
         RelativeArchivePath getParent() {
             if (!segments || segments.length == 1) {
                 return null
-            } else {
-                //Parent is always a directory so add / to the end of the path
-                String path = segments[0..-2].join('/') + '/'
-                return new RelativeArchivePath(setArchiveTimes(new ZipEntry(path)))
             }
+            //Parent is always a directory so add / to the end of the path
+            String path = segments[0..-2].join('/') + '/'
+            return new RelativeArchivePath(setArchiveTimes(new ZipEntry(path)))
         }
     }
 
@@ -482,6 +480,7 @@ class ShadowCopyAction implements CopyAction {
             return archivePath.classFile
         }
 
+        @SuppressWarnings(['GetterMethodCouldBeProperty'])
         @Override
         File getFile() {
             return null

--- a/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/internal/JRubyJarCopyAction.groovy
+++ b/jar-plugin/src/main/groovy/com/github/jrubygradle/jar/internal/JRubyJarCopyAction.groovy
@@ -55,9 +55,8 @@ import java.util.zip.ZipException
  * support the same behavior in a less hackish way.
  */
 @Slf4j
-@SuppressWarnings(['ParameterCount', 'CatchException', 'DuplicateStringLiteral',
-'CatchThrowable', 'VariableName'])
-class JRubyJarCopyAction implements CopyAction {
+class ShadowCopyAction implements CopyAction {
+    static final long CONSTANT_TIME_FOR_ZIP_ENTRIES = (new GregorianCalendar(1980, 1, 1, 0, 0, 0)).getTimeInMillis()
 
     private final File zipFile
     private final ZipCompressor compressor
@@ -65,9 +64,17 @@ class JRubyJarCopyAction implements CopyAction {
     private final List<Transformer> transformers
     private final List<Relocator> relocators
     private final PatternSet patternSet
+    private final ShadowStats stats
+    private final String encoding
+    private final GradleVersionUtil versionUtil
+    private final boolean preserveFileTimestamps
+    private final boolean minimizeJar
+    private final UnusedTracker unusedTracker
 
-    JRubyJarCopyAction(File zipFile, ZipCompressor compressor, DocumentationRegistry documentationRegistry,
-                            List<Transformer> transformers, List<Relocator> relocators, PatternSet patternSet) {
+    ShadowCopyAction(File zipFile, ZipCompressor compressor, DocumentationRegistry documentationRegistry,
+                            String encoding, List<Transformer> transformers, List<Relocator> relocators,
+                            PatternSet patternSet, ShadowStats stats, GradleVersionUtil util,
+                            boolean preserveFileTimestamps, boolean minimizeJar, UnusedTracker unusedTracker) {
 
         this.zipFile = zipFile
         this.compressor = compressor
@@ -75,10 +82,33 @@ class JRubyJarCopyAction implements CopyAction {
         this.transformers = transformers
         this.relocators = relocators
         this.patternSet = patternSet
+        this.stats = stats
+        this.encoding = encoding
+        this.versionUtil = util
+        this.preserveFileTimestamps = preserveFileTimestamps
+        this.minimizeJar = minimizeJar
+        this.unusedTracker = unusedTracker
     }
 
     @Override
     WorkResult execute(CopyActionProcessingStream stream) {
+        Set<String> unusedClasses
+        if (minimizeJar) {
+            stream.process(new BaseStreamAction() {
+                @Override
+                void visitFile(FileCopyDetails fileDetails) {
+                    // All project sources are already present, we just need
+                    // to deal with JAR dependencies.
+                    if (isArchive(fileDetails)) {
+                        unusedTracker.addDependency(fileDetails.file)
+                    }
+                }
+            })
+            unusedClasses = unusedTracker.findUnused()
+        } else {
+            unusedClasses = Collections.emptySet()
+        }
+
         final ZipOutputStream zipOutStr
 
         try {
@@ -91,7 +121,8 @@ class JRubyJarCopyAction implements CopyAction {
             withResource(zipOutStr, new Action<ZipOutputStream>() {
                 void execute(ZipOutputStream outputStream) {
                     try {
-                        stream.process(new StreamAction(outputStream, transformers, relocators, patternSet))
+                        stream.process(new StreamAction(outputStream, encoding, transformers, relocators, patternSet,
+                                unusedClasses, stats))
                         processTransformers(outputStream)
                     } catch (Exception e) {
                         log.error('ex', e)
@@ -103,27 +134,37 @@ class JRubyJarCopyAction implements CopyAction {
         } catch (UncheckedIOException e) {
             if (e.cause instanceof Zip64RequiredException) {
                 throw new Zip64RequiredException(
-                        String.format('%s\n\nTo build this archive, please enable the zip64 extension.\nSee: %s',
-                                e.cause.message, documentationRegistry.getDslRefForProperty(Zip, 'zip64'))
+                        String.format("%s\n\nTo build this archive, please enable the zip64 extension.\nSee: %s",
+                                e.cause.message, documentationRegistry.getDslRefForProperty(Zip, "zip64"))
                 )
             }
         }
-        return WorkResults.didWork(true)
+        return versionUtil.getWorkResult(true)
     }
 
-    private void processTransformers(ZipOutputStream s) {
+    private void processTransformers(ZipOutputStream stream) {
         transformers.each { Transformer transformer ->
             if (transformer.hasTransformedResource()) {
-                transformer.modifyOutputStream(s)
+                transformer.modifyOutputStream(stream, preserveFileTimestamps)
             }
         }
     }
 
-    @SuppressWarnings('EmptyCatchBlock')
+    private long getArchiveTimeFor(long timestamp) {
+        return preserveFileTimestamps ? timestamp : CONSTANT_TIME_FOR_ZIP_ENTRIES
+    }
+
+    private ZipEntry setArchiveTimes(ZipEntry zipEntry) {
+        if (!preserveFileTimestamps) {
+            zipEntry.setTime(CONSTANT_TIME_FOR_ZIP_ENTRIES)
+        }
+        return zipEntry
+    }
+
     private static <T extends Closeable> void withResource(T resource, Action<? super T> action) {
         try {
             action.execute(resource)
-        } catch (Throwable t) {
+        } catch(Throwable t) {
             try {
                 resource.close()
             } catch (IOException e) {
@@ -139,25 +180,16 @@ class JRubyJarCopyAction implements CopyAction {
         }
     }
 
-    class StreamAction implements CopyActionProcessingStreamAction {
-
-        private final ZipOutputStream zipOutStr
-        private final List<Transformer> transformers
-        private final List<Relocator> relocators
-        private final RelocatorRemapper remapper
-        private final PatternSet patternSet
-
-        private final Set<String> visitedFiles = [] as Set
-
-        StreamAction(ZipOutputStream zipOutStr, List<Transformer> transformers, List<Relocator> relocators,
-                            PatternSet patternSet) {
-            this.zipOutStr = zipOutStr
-            this.transformers = transformers
-            this.relocators = relocators
-            this.remapper = new RelocatorRemapper(relocators)
-            this.patternSet = patternSet
+    abstract class BaseStreamAction implements CopyActionProcessingStreamAction {
+        protected boolean isArchive(FileCopyDetails fileDetails) {
+            return fileDetails.relativePath.pathString.endsWith('.jar')
         }
 
+        protected boolean isClass(FileCopyDetails fileDetails) {
+            return FilenameUtils.getExtension(fileDetails.path) == 'class'
+        }
+
+        @Override
         void processFile(FileCopyDetailsInternal details) {
             if (details.directory) {
                 visitDir(details)
@@ -166,39 +198,112 @@ class JRubyJarCopyAction implements CopyAction {
             }
         }
 
+        protected void visitDir(FileCopyDetails dirDetails) {}
+
+        protected abstract void visitFile(FileCopyDetails fileDetails)
+    }
+
+    private class StreamAction extends BaseStreamAction {
+
+        private final ZipOutputStream zipOutStr
+        private final List<Transformer> transformers
+        private final List<Relocator> relocators
+        private final RelocatorRemapper remapper
+        private final PatternSet patternSet
+        private final Set<String> unused
+        private final ShadowStats stats
+
+        private Set<String> visitedFiles = new HashSet<String>()
+
+        StreamAction(ZipOutputStream zipOutStr, String encoding, List<Transformer> transformers,
+                            List<Relocator> relocators, PatternSet patternSet, Set<String> unused,
+                            ShadowStats stats) {
+            this.zipOutStr = zipOutStr
+            this.transformers = transformers
+            this.relocators = relocators
+            this.remapper = new RelocatorRemapper(relocators, stats)
+            this.patternSet = patternSet
+            this.unused = unused
+            this.stats = stats
+            if(encoding != null) {
+                this.zipOutStr.setEncoding(encoding)
+            }
+        }
+
         private boolean recordVisit(RelativePath path) {
             return visitedFiles.add(path.pathString)
         }
 
-        private void visitFile(FileCopyDetails fileDetails) {
-            try {
-                boolean isClass = (FilenameUtils.getExtension(fileDetails.path) == 'class')
-                if (!remapper.hasRelocators() || !isClass) {
-                    if (isTransformable(fileDetails)) {
-                        transform(fileDetails)
+        @Override
+        void visitFile(FileCopyDetails fileDetails) {
+            if (!isArchive(fileDetails)) {
+                try {
+                    boolean isClass = isClass(fileDetails)
+                    if (!remapper.hasRelocators() || !isClass) {
+                        if (!isTransformable(fileDetails)) {
+                            String mappedPath = remapper.map(fileDetails.relativePath.pathString)
+                            ZipEntry archiveEntry = new ZipEntry(mappedPath)
+                            archiveEntry.setTime(getArchiveTimeFor(fileDetails.lastModified))
+                            archiveEntry.unixMode = (UnixStat.FILE_FLAG | fileDetails.mode)
+                            zipOutStr.putNextEntry(archiveEntry)
+                            fileDetails.copyTo(zipOutStr)
+                            zipOutStr.closeEntry()
+                        } else {
+                            transform(fileDetails)
+                        }
+                    } else if (isClass && !isUnused(fileDetails.path)) {
+                        remapClass(fileDetails)
                     }
-                    else {
-                        String mappedPath = remapper.map(fileDetails.relativePath.pathString)
-                        ZipEntry archiveEntry = new ZipEntry(mappedPath)
-                        archiveEntry.setTime(fileDetails.lastModified)
-                        archiveEntry.unixMode = (UnixStat.FILE_FLAG | fileDetails.mode)
-                        zipOutStr.putNextEntry(archiveEntry)
-                        fileDetails.copyTo(zipOutStr)
-                        zipOutStr.closeEntry()
-                    }
-                } else if (isClass) {
-                    remapClass(fileDetails)
+                    recordVisit(fileDetails.relativePath)
+                } catch (Exception e) {
+                    throw new GradleException(String.format("Could not add %s to ZIP '%s'.", fileDetails, zipFile), e)
                 }
-                recordVisit(fileDetails.relativePath)
-            } catch (Exception e) {
-                throw new GradleException(String.format('Could not add %s to ZIP \'%s\'.', fileDetails, zipFile), e)
+            } else {
+                processArchive(fileDetails)
             }
+        }
+
+        private void processArchive(FileCopyDetails fileDetails) {
+            stats.startJar()
+            ZipFile archive = new ZipFile(fileDetails.file)
+            try {
+                List<ArchiveFileTreeElement> archiveElements = archive.entries.collect {
+                    new ArchiveFileTreeElement(new RelativeArchivePath(it))
+                }
+                Spec<FileTreeElement> patternSpec = patternSet.getAsSpec()
+                List<ArchiveFileTreeElement> filteredArchiveElements = archiveElements.findAll { ArchiveFileTreeElement archiveElement ->
+                    patternSpec.isSatisfiedBy(archiveElement.asFileTreeElement())
+                }
+                filteredArchiveElements.each { ArchiveFileTreeElement archiveElement ->
+                    if (archiveElement.relativePath.file) {
+                        visitArchiveFile(archiveElement, archive)
+                    }
+                }
+            } finally {
+                archive.close()
+            }
+            stats.finishJar()
         }
 
         private void visitArchiveDirectory(RelativeArchivePath archiveDir) {
             if (recordVisit(archiveDir)) {
                 zipOutStr.putNextEntry(archiveDir.entry)
                 zipOutStr.closeEntry()
+            }
+        }
+
+        private void visitArchiveFile(ArchiveFileTreeElement archiveFile, ZipFile archive) {
+            def archiveFilePath = archiveFile.relativePath
+            if (archiveFile.classFile || !isTransformable(archiveFile)) {
+                if (recordVisit(archiveFilePath) && !isUnused(archiveFilePath.entry.name)) {
+                    if (!remapper.hasRelocators() || !archiveFile.classFile) {
+                        copyArchiveEntry(archiveFilePath, archive)
+                    } else {
+                        remapClass(archiveFilePath, archive)
+                    }
+                }
+            } else {
+                transform(archiveFile, archive)
             }
         }
 
@@ -211,20 +316,36 @@ class JRubyJarCopyAction implements CopyAction {
             }
         }
 
+        private boolean isUnused(String classPath) {
+            final String className = FilenameUtils.removeExtension(classPath)
+                    .replace('/' as char, '.' as char)
+            final boolean result = unused.contains(className)
+            if (result) {
+                log.debug("Dropping unused class: $className")
+            }
+            return result
+        }
+
         private void remapClass(RelativeArchivePath file, ZipFile archive) {
             if (file.classFile) {
-                addParentDirectories(new RelativeArchivePath(new ZipEntry(remapper.mapPath(file) + '.class'), null))
-                remapClass(archive.getInputStream(file.entry), file.pathString)
+                ZipEntry zipEntry = setArchiveTimes(new ZipEntry(remapper.mapPath(file) + '.class'))
+                addParentDirectories(new RelativeArchivePath(zipEntry))
+                InputStream is = archive.getInputStream(file.entry)
+                try {
+                    remapClass(is, file.pathString, file.entry.time)
+                } finally {
+                    is.close()
+                }
             }
         }
 
         private void remapClass(FileCopyDetails fileCopyDetails) {
             if (FilenameUtils.getExtension(fileCopyDetails.name) == 'class') {
-                remapClass(fileCopyDetails.file.newInputStream(), fileCopyDetails.path)
+                remapClass(fileCopyDetails.file.newInputStream(), fileCopyDetails.path, fileCopyDetails.lastModified)
             }
         }
 
-        private void remapClass(InputStream classInputStream, String path) {
+        private void remapClass(InputStream classInputStream, String path, long lastModified) {
             InputStream is = classInputStream
             ClassReader cr = new ClassReader(is)
 
@@ -240,7 +361,7 @@ class JRubyJarCopyAction implements CopyAction {
             try {
                 cr.accept(cv, ClassReader.EXPAND_FRAMES)
             } catch (Throwable ise) {
-                throw new GradleException('Error in ASM processing class ' + path, ise)
+                throw new GradleException("Error in ASM processing class " + path, ise)
             }
 
             byte[] renamedClass = cw.toByteArray()
@@ -248,42 +369,75 @@ class JRubyJarCopyAction implements CopyAction {
             // Need to take the .class off for remapping evaluation
             String mappedName = remapper.mapPath(path)
 
+            InputStream bis = new ByteArrayInputStream(renamedClass)
             try {
                 // Now we put it back on so the class file is written out with the right extension.
-                zipOutStr.putNextEntry(new ZipEntry(mappedName + '.class'))
-                IOUtils.copyLarge(new ByteArrayInputStream(renamedClass), zipOutStr)
+                ZipEntry archiveEntry = new ZipEntry(mappedName + ".class")
+                archiveEntry.setTime(getArchiveTimeFor(lastModified))
+                zipOutStr.putNextEntry(archiveEntry)
+                IOUtils.copyLarge(bis, zipOutStr)
                 zipOutStr.closeEntry()
             } catch (ZipException e) {
-                log.warn('We have a duplicate ' + mappedName + ' in source project')
+                log.warn("We have a duplicate " + mappedName + " in source project")
+            } finally {
+                bis.close()
             }
         }
 
-        private void visitDir(FileCopyDetails dirDetails) {
+        private void copyArchiveEntry(RelativeArchivePath archiveFile, ZipFile archive) {
+            String mappedPath = remapper.map(archiveFile.entry.name)
+            ZipEntry entry = new ZipEntry(mappedPath)
+            entry.setTime(getArchiveTimeFor(archiveFile.entry.time))
+            RelativeArchivePath mappedFile = new RelativeArchivePath(entry)
+            addParentDirectories(mappedFile)
+            zipOutStr.putNextEntry(mappedFile.entry)
+            InputStream is = archive.getInputStream(archiveFile.entry)
+            try {
+                IOUtils.copyLarge(is, zipOutStr)
+            } finally {
+                is.close()
+            }
+            zipOutStr.closeEntry()
+        }
+
+        @Override
+        protected void visitDir(FileCopyDetails dirDetails) {
             try {
                 // Trailing slash in name indicates that entry is a directory
                 String path = dirDetails.relativePath.pathString + '/'
                 ZipEntry archiveEntry = new ZipEntry(path)
-                archiveEntry.setTime(dirDetails.lastModified)
+                archiveEntry.setTime(getArchiveTimeFor(dirDetails.lastModified))
                 archiveEntry.unixMode = (UnixStat.DIR_FLAG | dirDetails.mode)
                 zipOutStr.putNextEntry(archiveEntry)
                 zipOutStr.closeEntry()
                 recordVisit(dirDetails.relativePath)
             } catch (Exception e) {
-                throw new GradleException(String.format('Could not add %s to ZIP \'%s\'.', dirDetails, zipFile), e)
+                throw new GradleException(String.format("Could not add %s to ZIP '%s'.", dirDetails, zipFile), e)
             }
         }
 
         private void transform(ArchiveFileTreeElement element, ZipFile archive) {
-            transform(element, archive.getInputStream(element.relativePath.entry))
+            transformAndClose(element, archive.getInputStream(element.relativePath.entry))
         }
 
         private void transform(FileCopyDetails details) {
-            transform(details, details.file.newInputStream())
+            transformAndClose(details, details.file.newInputStream())
         }
 
-        private void transform(FileTreeElement element, InputStream is) {
-            String mappedPath = remapper.map(element.relativePath.pathString)
-            transformers.find { it.canTransformResource(element) }.transform(mappedPath, is, relocators)
+        private void transformAndClose(FileTreeElement element, InputStream is) {
+            try {
+                String mappedPath = remapper.map(element.relativePath.pathString)
+                transformers.find { it.canTransformResource(element) }.transform(
+                        TransformerContext.builder()
+                                .path(mappedPath)
+                                .is(is)
+                                .relocators(relocators)
+                                .stats(stats)
+                                .build()
+                )
+            } finally {
+                is.close()
+            }
         }
 
         private boolean isTransformable(FileTreeElement element) {
@@ -295,12 +449,10 @@ class JRubyJarCopyAction implements CopyAction {
     class RelativeArchivePath extends RelativePath {
 
         ZipEntry entry
-        FileCopyDetails details
 
-        RelativeArchivePath(ZipEntry entry, FileCopyDetails fileDetails) {
+        RelativeArchivePath(ZipEntry entry) {
             super(!entry.directory, entry.name.split('/'))
             this.entry = entry
-            this.details = fileDetails
         }
 
         boolean isClassFile() {
@@ -310,15 +462,14 @@ class JRubyJarCopyAction implements CopyAction {
         RelativeArchivePath getParent() {
             if (!segments || segments.length == 1) {
                 return null
+            } else {
+                //Parent is always a directory so add / to the end of the path
+                String path = segments[0..-2].join('/') + '/'
+                return new RelativeArchivePath(setArchiveTimes(new ZipEntry(path)))
             }
-
-            //Parent is always a directory so add / to the end of the path
-            String path = segments[0..-2].join('/') + '/'
-            return new RelativeArchivePath(new ZipEntry(path), null)
         }
     }
 
-    @SuppressWarnings('GetterMethodCouldBeProperty')
     class ArchiveFileTreeElement implements FileTreeElement {
 
         private final RelativeArchivePath archivePath
@@ -384,6 +535,10 @@ class JRubyJarCopyAction implements CopyAction {
         @Override
         int getMode() {
             return archivePath.entry.unixMode
+        }
+
+        FileTreeElement asFileTreeElement() {
+            return new DefaultFileTreeElement(null, new RelativePath(!isDirectory(), archivePath.segments), null, null)
         }
     }
 }

